### DISCUSSION TITLE
Rule Proposal: Ban usage of experimental and developer preview features

### DISCUSSION THREAD

**Discussion by florianeichin** [angular-eslint/angular-eslint#2305] [New Rule Proposals]
We would love to see the linter complaining on angular developer preview features, as they should not be used in our plattform for stability reasons. Is there a chance to get that?

**Comment by json-derulo** [comment]
For all TypeScript functions, this could work somewhat reliably by checking the `@developerPreview` JSDoc tags, assuming that Angular consistently sets this JSDoc tag on their features which are in developer preview. This would work in a generic way. Though I can imagine this could be costly operations with performance implications.

However there might be functionality in developer preview which isn't used in TypeScript context, such as Control flow, experimental builders, experimental test runners, ... For this kind of functionality, IMHO it doesn't make sense to implement  such a rule, as it needs to be implemented and kept up-to-date manually.

Would you be fine with a solution which exclusively works for TypeScript?

**Comment by marleypowell** [comment]
> For all TypeScript functions, this could work somewhat reliably by checking the `@developerPreview` JSDoc tags, assuming that Angular consistently sets this JSDoc tag on their features which are in developer preview. This would work in a generic way. Though I can imagine this could be costly operations with performance implications.
> 
> However there might be functionality in developer preview which isn't used in TypeScript context, such as Control flow, experimental builders, experimental test runners, ... For this kind of functionality, IMHO it doesn't make sense to implement such a rule, as it needs to be implemented and kept up-to-date manually.
> 
> Would you be fine with a solution which exclusively works for TypeScript?

Exclusively Typescript would be better than nothing but there's a handful of new template features that are in preview,

It's a shame that it can't be a compiler option - https://github.com/angular/angular/issues/57656

**Comment by NateRadebaugh** [comment]
I stumbled upon this old discussion. Looks like these have been implemented in this plugin:
* `@angular-eslint/no-developer-preview` - [docs/rules/no-developer-preview.md](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/no-developer-preview.md)
* `@angular-eslint/no-experimental` - [docs/rules/no-experimental.md](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/no-experimental.md)

**Comment by JamesHenry** [comment]
As noted, the rules exist
